### ISSUE TITLE
SAF migration: Better user experience

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#9872]
Taken from https://github.com/cgeo/cgeo/issues/9606#issuecomment-767478209 by @MagpieFourtyTwo

Anyway - I know, it's hard to explain what happens behind the scenes, but to me this text and the steps one has to take afterwards, are way to much to handle for Joe Average. In order to not getting the support flooded with questions about which directories to chose, how to get back maps and stuff and where will my exports go .., we need a clear list of steps over there, which tells the user how he can get to the situation he had before he was forced to grant access to some folders ...

---

I wonder a bit about your confusion here. AFAIR from my pre-tests of the SAF nightlies the usage is pretty straight forward:
- Nothing will ever be lost also if you select another than the existing `/cgeo` as base dir (because old data either remains or gets copied/moved)
- Offline map directory should IMHO be kept as it was before upgrading to SAF, I might be wrong though @eddiemuc 
- Upon exporting something from c:geo you will be informed where it was stored (same as before)

**Comment by fm-sys** [comment]
> I wonder a bit about your confusion here. AFAIR from my pre-tests of the SAF nightlies the usage is pretty straight forward

I think the biggest mistake we make is that we overestimate our average users. The only thing they want is that everything works without problems. As has been said several times, the copy/move selection options are avoidable and not clear enough to non IT experts. Also the "explanatory text" is way too long. Users don't want to be told big why something is technically necessary, they just want to know what they need to do - with words as simple and compact as possible

**Comment by MagpieFourtyTwo** [comment]
After closing the message window, I got exactly one more dialog to chose a folder and I selected ``internal\cgeo``. as has been recommended before. Main reason for my "bad feeling": I have had no clue at all, which settings have been effected by this choice. Main effect was, however, that there were no maps anymore in my Map selection drop down list - and this was because the above procedure most obviously changed my mapsVector folder from ``sdcard\Locus\mapsVector`` to ``internal\cgeo\maps``.

Have not said that something has been deleted, cause after switching the folder back the map list was there again, of course.

Nevertheless, kind of irritating, that in the Map Data config section the ``Directory with offline maps`` is shown as ``...\Locus\mapsVector (User-defined)``, no matter if I have chosen the folder from SD card or Internal Storage.

**Comment by MagpieFourtyTwo** [comment]
> I think the biggest mistake we make is that we overestimate our average users. The only thing they want is that everything works without problems. As has been said several times, the copy/move selection options are avoidable and not clear enough to non IT experts. Also the "explanatory text" is way too long. Users don't want to be told big why something is technically necessary, they just want to know what they need to do - with words as simple and compact as possible

Ex-act-ly! Usually I would just click ``thumbs up``, but this paragraph is *way* too true, so I just had to do a full quote ...  ;)

**Comment by MagpieFourtyTwo** [comment]
> Nevertheless, kind of irritating, that in the Map Data config section the ``Directory with offline maps`` is shown as ``...\Locus\mapsVector (User-defined)``, no matter if I have chosen the folder from SD card or Internal Storage.

Ok, just had a look at the settings list (via ``System`` => ``View settings``) - obviously the "real" path is a little bit too clumsy to show it in a user dialog. ;) Understood.

However, the string of ``persistablefolder_offlinemaps`` contains ``.../tree/0123-4567%3ALocus`` for SD card vs. ``tree/primary%3ALocus`` for internal memory ... thus I would suggest to either show the stuff after the ``/tree/`` part or explicitly add ``SD card`` or ``internal`` depending on occurrence of either ``primary`` or ```123-4567`` ...

**Comment by MagpieFourtyTwo** [comment]
BTW: Same applies for ``Settings`` => ``System`` => ``Local File System`` => ``Base directory`` - and most probably for all following directory settings.  ;)

**Comment by eddiemuc** [comment]
This is very valuable input, thanks to all of it!

To be clear: I never intended or expected that the SAF change UX would go to production as it is right now. I was eager for this UX conversation to clarify exactly those questions. You should have an impression from the current state on what is possible (and I can start fixing errors :-)).

* Regarding the Uri display format: in an initial version I displayed /tree/primary or whatever was there instead of the ... Was changed upon request. Still discussable. Also compare to e.g. Signal implementation. Note that unfortunately Uri formats can be different, even for same folder, and also across devices and Android versions...

* offline folder change is a bug, should not happen. 

* We already discussed integration of all of this into the installation wizard. Now would be the time to do it

**Comment by Lineflyer** [comment]
> However, the string of persistablefolder_offlinemaps contains .../tree/0123-4567%3ALocus for SD card vs. tree/primary%3ALocus for internal memory ... thus I would suggest to either show all the stuff after the /tree/ part or explicitly add SD card or internal depending on occurrence of either primary or 123-4567 ...

Yep, that primarily why I opened #9809 (as I found an app which managed to show it better on some occasions). But it turned out to be also just clever guessing. Its somehow a strange implementation by Android :(

**Comment by eddiemuc** [comment]
@Lineflyer @MagpieFourtyTwo @fm-sys @geocachermgu @moving-bits 

Now the most prominent bugs are fixed (hopefully...) and if I understand it correctly saf migration is Integrated initially in the wizard. It is time to finalize how UX should be.

I am not the right person to decide on this, my UX will always be too complicated. So I need you to help me out here:

* Should the user have the option on initial folder select between donothing/copy/move or should we make this decision for him?
* If we make this decision: which should it be?
* If not: how can this be made clearer to him UX-wise

* Should we leave donothing/copy/move choice on the normal folder select? (E.g. offline maps). If not: shall there be a default choice, and which one?

* Should we leave the user the choice to select "default" folder (which also allows him to go back to the legacy base folder if he has already migrated to a new one). Or should we simply delete this option? Note that in this case we will not have an option to help users in support in case something goes wrong in the field, so we need to test this stuff very throughly!

* Any other suggestions wrt UX

@fm-sys if in the questions above it turns out that copy/move is still needed I will polish it up with progress bar, better error Handling, cancelation option etc. But I want to wait for UX discussion to end before - if copy/move is no longer needed I can spare the work.

**Comment by moving-bits** [comment]
> Should the user have the option on initial folder select between donothing/copy/move or should we make this decision for him?

For initial migration I would make the decision for them, and probably "move" is the action needed in this use-case. (Can we auto-detect whether old and new folder are the same, and skip moving in this case?)

But I would leave a "cancel" option in the dialog, so that users not having time at that moment (or testers that want to test specific constellations) can skip the action (which should reappear on next start of c:geo).

> Should we leave donothing/copy/move choice on the normal folder select? (E.g. offline maps). If not: shall there be a default choice, and which one?

Why should a regular user need a "do nothing" option at all (if they have a "cancel" option in the dialog)?
Also I'm unsure about the copy option. I tend to think, an implicit move would be enough (which auto-detection of old folder = new folder).

> Should we leave the user the choice to select "default" folder (which also allows him to go back to the legacy base folder if he has already migrated to a new one).

Hm, "default" may be misleading. I would have expected "default" to be a c:geo standard location and folder structure, but still SAF access. This should be the preset on folder selection, when the stored legacy location is not longer valid or (on fresh install) does not exist.

To be able to resort to a legacy location might be an extra option, but only in case of a migrated system, not for fresh installs (as there are no legacy folders).

> Or should we simply delete this option? Note that in this case we will not have an option to help users in support in case something goes wrong in the field, so we need to test this stuff very throughly!

At least in the beginning we should keep us that option. If everything works well for some time, we might remove that option.

> Any other suggestions wrt UX

Keep it as simple as possible for the standard user :-)
If we need some sort of expert mode for testers or for our support, we might put this into a menu.

> I will polish it up with progress bar, better error Handling, cancelation option etc.

Sounds good!

Thanks @eddiemuc for all the work you're putting into this and pursuing the task to make the transition so smooth for all the users, testers, support and dev team!

**Comment by Lineflyer** [comment]
> Should the user have the option on initial folder select between donothing/copy/move or should we make this decision for him?
>  If we make this decision: which should it be?

For migration it might be wise just trying to move everything and report if it was OK or not.


> Should we leave donothing/copy/move choice on the normal folder select? (E.g. offline maps). If not: shall there be a default choice, and which one?

For later change of the base folder as well as folder selection for offline maps we should still provide the donothing/copy/move dialog. 

Summarized it probably means: If the current folder is [Legacy] we could move automatically, if the current folder is [SAF] we should offer the dialog to copy/move/do nothing.

> Should we leave the user the choice to select "default" folder (which also allows him to go back to the legacy base folder if he has already migrated to a new one). Or should we simply delete this option? Note that in this case we will not have an option to help users in support in case something goes wrong in the field, so we need to test this stuff very throughly!

We should IMHO not allow to return to the legacy working mode.  We have time for testing now.

> Any other suggestions wrt UX

Need to check the enhanced nstallation wizard and will try to doublecheck some migration cases on different devices and API levels (depends on my available devices).

@MagpieFourtyTwo told, that the initial description is not understandable. To be honest: I think its very well explained and as easy as possible. The only other thing I could imagine is, to leave it with "You have to select your base folder. If you want more details click here: [link to FAQ]". 
But IMHO we can leave it like it is now. If the user is unsure, he can still (at any point) abort the process and the dialog will come up again at next start, so he can again read and understand it.

**Comment by eddiemuc** [comment]
I will wait some more time to collect feedback. In the meantime to clarify some things:

* The copy/move function already handles the case of "source and target folder are same" (in which case nothing is done), along with the more elaborate cases of " target folder is a subfolder of source folder" and vice versa. Was not so easy to implement. Remember that with Uris, two very different looking Uris May point to same folder! (Had to work with marker files)

* "Donothing" above refers only to the copy/move part of the folder selection. If user chooses "donothing" today, folder will still be changed but no attempt happens to copy/move any content

From the discussion I conclude at least that copy/move in one way or the other is still needed. Thus I take it as next task to make this more elaborate (progress bar etc). This should be enough to do for me for next couple of days :-)

Keep on the good discussion and thanks a lot for the good input!

**Comment by MagpieFourtyTwo** [comment]
> @MagpieFourtyTwo told, that the initial description is not understandable.

I in fact said, it is not understandable *for average users.* 

> To be honest: I think its very well explained and as easy as possible.

The  problem is - and I had another perfect example today, when a cachemate stumbled upon the message on start of c:geo - it's too much information in one screen and too little explicit instructions, what must be done on next step.

> The only other thing I could imagine is, to leave it with "You have to select your base folder. If you want more details click here: [link to FAQ]".
> But IMHO we can leave it like it is now. If the user is unsure, he can still (at any point) abort the process and the dialog will come up again at next start, so he can again read and understand it.

May possibly sound a bit harsh, but, to be honest, all this stuff is completely superfluous, IMO. Average users do not care at all for internal needs. All they want is an app, which just continues to work as usual. And to achieve this esp. in case of a migration, is as easy as one, two, three ... cause in most of the cases, if not in all of them, at the end everything should be as it was before, anyway. In fact, as the milkmaid would say: "just the folder names change, nothing else, so what's all the fuss about?!

Especially in case of migration, users have their directories set up, already. And they have/had reasons to do it right the way it is now. So most probably, they just want to keep things as they are. And stuff where it is right now. Cause this is, what they know. Or (most probably even more often) someone else has set up for them ... so, no need for copying, no need to move, nothing. just preserve it as it is. That's the usual wish. And, esp. for maps, the only way, to let e. g. Locus and c:geo still live happily together without wasting valuable memory space.

So why not just ease the process and just ask for granting access to the already configured directories? Like ...
1. Introducing preamble: "Due to upcoming changes in Android, c:geo needs your renewed explicit approval to access data on your phone. Therefore in the following dialogs, please grant access to c:geo for the folders you currently have configured:" Then ...
2. one dialog for data like backups and stuff, starting with current dir, e. g. ``<internal>\cgeo``
3. one dialog for map files, starting in the currently configured map directory
4. one dialog for themes, starting in the currently configured directory
5. one dialog for GPX ex- and imports ...

Important: One by one. And every single one explained. With a concrete naming of the currently set value and the way to go/navigate there ... and afterwards a short comparison before / after, so the user knows, what has happened.

Ok, maybe I overlook some complexity, but esp this complexity must be made completely transparent for the average user. Otherwise we will get tons of non-working c:geo incarnations ... for sure. I just have to remember the completely disenchanted facial expression at the beginning of today's caching tour ... :/

**Comment by eddiemuc** [comment]
> So why not just ease the process and just ask for granting access to the already configured directories? Like ...

> one dialog for data like backups and stuff, starting with current dir, e. g. <internal>\cgeo

Unfortunately it is technically not possible to force the select intent to start with the previous folder if it is not already a "migrated" one (the intent seems to ignore File-based Uris). So we can't prevent the selection to start with another folder (often it is the download folder or the last one selected in any other context/app with this intent)

> one dialog for GPX ex- and imports ...

We agreed to reduce this to one dir for both export and import

In fact, my thinking is the same as yours, only restricted by de facto technical limitationals. I included the whole copy stuff on explicit wish, but it is perfectly fine to hide it.

The selection of offline map folder, GPX folder etc I wouldn't even include in normal users wizard. This imho is already advanced stuff, and users who have set these options should be capable of reselecting them on their own in the settings.

**Comment by MagpieFourtyTwo** [comment]
> Unfortunately it is technically not possible to force the select intent to start with the previous folder if it is not already a "migrated" one

Ok, already was afraid of s/th like this.

Problem is: Most of the c:geo users I know have *not* set up their offline maps and stuff on their own, thus the assumption ...
> users who have set these options should be capable of reselecting them on their own in the settings

... is in fact valid for advanced users only. But, most regretfully, does not meet reality for the majority of the c:geo users (I know). Most of these were even overexerted by setting up their map folder (esp. with the old selection dialog).

Thus c:geo should better take them by the hand and a) first show the current setting, so they can at least get an impression of the aim we're heading for, and then b) give an explicit advice how to achieve a clean switch.

And this, as already said, step by step, with as less complexity per step as possible, even if this may result in a dialog with some more windows, cause things are way faster clicked than read. And if there's too much text in one go, they just click without reading anyway.

Thus there should first be the explaining preamble mentioned above, then next screen with current setting and recommendation for the first folder, then second ... in fact perhaps s/th like ...
- ``Choose folder for cgeo data (for backups etc.)``
   - Current setting: ``/storage/0/cgeo`` (while ``.*/0/`` appears to be pretty common for ``internal storage``, thus should perhaps be replaced by ``internal memory`` for better understanding)
   - Recommendation: go to ``Internal Memory`` => ``cgeo`` and grant access
- ``Choose folder with your Map files``
   - Current setting: ``/storage/1234-567/Locus/mapsVector`` (while ``.*/[0-9]+-[0-9]+/`` appears to be pretty common for ``external storage`` or ``SD Card``, thus should perhaps be replaced by ``SD card`` for better understanding)
   - Recommendation: go to ``SD Card`` => ``Locus`` => ``mapsVector`` and grant access
- ...

and every single folder config with an own window, step by step - [KISS](https://en.wikipedia.org/wiki/KISS_principle)! Otherwise, I am pretty sure, support will be flooded, and perhaps we will even lose installations, cause users will end up without offline maps, missing themes, not finding their backups or being unable to use Locus anymore, cause c:geo moved their map files away ...

BTW: Ex- and Import can of course be handled in one step, that's why I put them in just one line.  ;)

**Comment by MagpieFourtyTwo** [comment]
BTW: Things get even harder with the current situation, as such questions usually arose (and could successfully be solved) while being on events ...

**Comment by fm-sys** [comment]
> > So why not just ease the process and just ask for granting access to the already configured directories? Like ...
> 
> 
> Unfortunately it is technically not possible to force the select intent to start with the previous folder if it is not already a "migrated" one (the intent seems to ignore File-based Uris)

Are you sure that there isn't a way to convert a file URI to a tree URI. I've read something like this on SO, even if it's a little bit hacky. Still it would worth to include such a workaround! 

I will play around with it a bit and create a PR if it works...

**Comment by eddiemuc** [comment]
> Are you sure that there isn't a way to convert a file URI to a tree URI. I've read something like this on SO, even if it's a little bit hacky. Still it would worth to include such a workaround!
> 
> I will play around with it a bit and create a PR if it works...

I am not sure of anything in this area :-) Just haven't found a way yet.

The parameter to set to open a document tree select intent on a defined place is `DocumentsContract.EXTRA_INITIAL_URI`. In our code I construct this intent and set this parameter in method `ContentStorageActivityHelper.selectFolderInternal`.

Of course you can create an Uri out of a File, those have then "file://" in front of them. If a Folder is a file folder, then the above parameter will indeed be set to such a file Uri. However, at least in my tests these Uris seem to get ignored by the intent dialog ("content://" Uris work). 

But I haven't found a generic way to construct a working Tree Uri ("content://...") out of a File which then points to the same folder. If this is somehow possible I would be glad to learn how. 

(To prevent you from falling into a "trap": `DocumentFile `will not help you here. It will just encapsulate the "file://" Uri, not create a "content://" Uri out of it)

**Comment by Lineflyer** [comment]
Side note: While investigating the possibility to improve the process please keep in mind that these functions have to work down to API 21.

**Comment by moving-bits** [comment]
Thanks, guys, for this discussion and pursuit for a good balance of the different aspects of user-friendlyness.

One more question in this area from my side: What is the right point in time to ask for the different folder grants?

I see several options:
1. Only ask for base folder on install/migration, and ask all the other folders on-demand.
2. Do them all during first-time configuration, be it a fresh install or a migration.
3. Like 1), but on migration also ask for permissions for all folders currently set. (So if a user has never set a gpx import folder, they won't get asked for it on migration. But if they have set an offline map folder, it get's included during migration.)

IMHO 3) is a good compromise of not asking too many questions, but also having c:geo up and ready as fast as possible.

**Comment by fm-sys** [comment]
As the old legacy folders will still be accessible until the user uninstalls c:geo (preverseLegacyExternalStorage flag which will still work on SDK 30), there is IMHO no need to ask for granting access at all. Only as soon as they can't be accessed anymore c:geo should ask for permissions...

**Comment by fm-sys** [comment]
> Whatever we agree upon regarding this question should be added to the installation wizard, IMHO.

BTW, currently the wizard opens with "welcome to c:geo" when migrating from a previous version which is quite confusing. Also setting up "Services" and "Advanced configuration" should be skipped. Otherwise, users might think that all their data was deleted...

**Comment by moving-bits** [comment]
Ok, maybe we need to distinguish between "migration" and "first installation" modes in the wizard, and skip some part when in "migration" mode.

Although I would leave the advanced settings in there even in "migration" mode, just to make a bit advertisement for BRouter etc. :-) What do you think?

**Comment by moving-bits** [comment]
> As the old legacy folders will still be accessible until the user uninstalls c:geo (preverseLegacyExternalStorage flag which will still work on SDK 30), there is IMHO no need to ask for granting access at all. Only as soon as they can't be accessed anymore c:geo should ask for permissions...

I tend to disagree on that, but I'm indecisive yet: The user needs to do the migration anyway in some point of time, so I tend to do it the earlier the better to have some room/access for follow-up if needed. If we spare it until the latest possible time, and for whatever reason some very specific situation fails, we are not able to help the user.

**Comment by eddiemuc** [comment]
> As the old legacy folders will still be accessible until the user uninstalls c:geo (preverseLegacyExternalStorage flag which will still work on SDK 30), there is IMHO no need to ask for granting access at all. Only as soon as they can't be accessed anymore c:geo should ask for permissions...

Please note: when cgeo first notices that it can't access the original folders any more it will at the same time be too late for cgeo to move the old folders content to the new one on change (because cgeo can't access the old folders content an more)

**Comment by fm-sys** [comment]
> Although I would leave the advanced settings in there even in "migration" mode, just to make a bit advertisement for BRouter etc. :-) What do you think?

![grafik](https://user-images.githubusercontent.com/64581222/106516074-da1c5180-64d6-11eb-96dc-a3b149cae59c.png)

Do you think that the already existing advertising is not enough? ;-)
I would keep the "migration" as short as possible, as every additional step can potentially confuse users...

**Comment by moving-bits** [comment]
@eddiemuc 
Is there a method to reliably detect whether migration for a certain `PersistableFolder` has been done already? (I guess it's not the same as with BASE folder, as others may be set to default.) This would be sort of a prerequisite for the wizard to also handle the migration case for those folders.

**Comment by eddiemuc** [comment]
> @eddiemuc 
> Is there a method to reliably detect whether migration for a certain `PersistableFolder` has been done already? (I guess it's not the same as with BASE folder, as others may be set to default.) This would be sort of a prerequisite for the wizard to also handle the migration case for those folders.

Not yet, but can be build easily in Settings. See the legacy methods there, I prefix each migrated value with a constant string. Whenever this is set, value can be considered migrated.

**Comment by MagpieFourtyTwo** [comment]
> Is there a method to reliably detect whether migration for a certain ``PersistableFolder`` has been done already?

Milkmaid's saying: Why not just guess?
While surfing through the list of settings the values for old and new version look completely different,
so it should be not too hard, to distinguish between them ...?
Or am I resp. is this just another proof of [Mark Twains remark (if it was his at all)?](https://quoteinvestigator.com/2016/07/17/solution)  ;)

**Comment by eddiemuc** [comment]
With the integration of migration into the wizard and the simplification of the process that came with it (many thanks to @moving-bits): can this issue be closed now, or are there any open aspects left?

**Comment by Lineflyer** [comment]
Lets close for now and (in case of need) open dedicated issue for remaining suggestions.
I will carefully test the relevant migration cases once again next week if time permits.

**Comment by fm-sys** [comment]
> can this issue be closed now, or are there any open aspects left?

There is still quite a bit to do. Further up, @eddiemuc had wrote down following list:

> It is time to finalize how UX should be.
> 
> 
>     * Should the user have the option on initial folder select between donothing/copy/move or should we make this decision for him?
>     * If we make this decision: which should it be?

IIRC we had decided to always move the content on SAF migration. So no option for the user to choose manually. 

>     * Should we leave donothing/copy/move choice on the normal folder select? (E.g. offline maps). If not: shall there be a default choice, and which one?

Here I'm undecided. Are there any reasons why the content should not always be moved as well? 
>     * Should we leave the user the choice to select "default" folder (which also allows him to go back to the legacy base folder if he has already migrated to a new one). Or should we simply delete this option?

Oh, I thought  "default" refers to the default sub-folder in the c:geo base dir, which would be a quite nice feature for resetting the folder path.

---

Should we reopen the issue or rather create a new one with this "todo summary" for future discussion/implementation?

**Comment by eddiemuc** [comment]
For the first question: @moving-bits has implented the wizard such that for migration there are no questions asked, content is just moved

For the second question: from my personal perspective as a user, I insist of keeping the option "do nothing" which I will always use myself. I want to be in full control of my public files and not relying on what the app decide is best. The rest I don't have any opinion, whatever suites you best.

For the third option: what @fm-sys says can be done with this option too. I had the feeling that it should be kept, but I am not insisting in anything.

**Comment by eddiemuc** [comment]
One additional comment: it was a lot (!) of work to create a full-featured copy/move function, and all together I surely created several hundred lines of code for this. It would be a bit of a bummer for me if this function would just be one-time-used for migration and then never again.

**Comment by fm-sys** [comment]
> One additional comment: it was a lot (!) of work to create a full-featured copy/move function, and all together I surely created several hundred lines of code for this. It would be a bit of a bummer for me if this function would just be one-time-used for migration and then never again.

Don't worry. The copy/move feature is great. The only question is, if we should always use it or if the user can decide by his self.

> For the first question: @moving-bits has implented the wizard such that for migration there are no questions asked, content is just moved

Thanks for the hint. Seems like I missed this chance. Tested it, and everything works fine. Really nice done and IMHO easy to understand from user perspective.

There is only one little bug left. "0/50 files progressed" doesn't sound quite successful, while actually everything was done (as I selected the old folder and nothing needed to be copied). I will create a PR for this one... 
![2021-02-06 23_51_58-cgeo Android Studio](https://user-images.githubusercontent.com/64581222/107131495-492ee700-68d7-11eb-9a8a-4797c539dc64.png)


> For the third option ("default button"): what @fm-sys says can be done with this option too. I had the feeling that it should be kept, but I am not insisting in anything.

Ok. Let's keep it for migration tests / fixing upgrade problems. We can change it later anytime in the future.

**Comment by Lineflyer** [comment]
> One additional comment: it was a lot (!) of work to create a full-featured copy/move function, and all together I surely created several hundred lines of code for this. It would be a bit of a bummer for me if this function would just be one-time-used for migration and then never again.

Just now reading this.
Are you aware, that there is a move function already implemented for the move of "Geocache data" between possible targets? Maybe these codes can be combined/unified (while that mentioned function does not involve SAF).

**Comment by eddiemuc** [comment]
> > One additional comment: it was a lot (!) of work to create a full-featured copy/move function, and all together I surely created several hundred lines of code for this. It would be a bit of a bummer for me if this function would just be one-time-used for migration and then never again.
> 
> Just now reading this.
> Are you aware, that there is a move function already implemented for the move of "Geocache data" between possible targets? Maybe these codes can be combined/unified (while that mentioned function does not involve SAF).

I am aware of this. This code can't be reused for several reasons:
* It uses the file interface. SAF works with Uris and Streams. This complicates things quite a bit
* The copy/move has also to work in mixed scenarios (copy file-folder to Saf-folder etc)
* There is also a copy-function provided by DocumentContract for Documents, but this function is only available with SDK24, also it does not work for mixed cases.
* It has no protection or checks against common special cases (source folder same as target folder, target folder is subdirs of source etc). Those special cases are not happening in the context this function is used, but they are happening in context of freely user-selected folders
* There is no support on current function to write a GUI with progress bar etc for longer running copies/moves. If I see it correctly then this would simply block the app (may run asynchronously, I didn't check that)
* Function is not unit-tested

**Comment by eddiemuc** [comment]
(when I say "copy" I also mean "move" of course. Technically, "move" is more complicated than "copy": move is copying plus deleting the source afterwards (but not always all of it eg when target of move is subdir of source ...)

**Comment by eddiemuc** [comment]
One additional thing came to my mind, you might want to discuss this too:

 when Implementing copy/move I had to make a decision what should happen when the file to copy/move already exists in target (more precise: a file with same name exists in the place): currently in this case the file to copy/move is renamed in the target by appending a number (e.g. (1)). This is the only way to make sure not to accidentally delete user content, other options are:
* Silently overwrite
* Ask user 

For the "ask user" option there would be several subquestions too, e.g. shall this be a "collective" question ("12 files exist in target shall they be overwritten, ignored or duplicated") or a question per file etc. Can get very complicated too :-)

@Lineflyer haven't checked what happens in this case for the File-based routine.

OK for everyone?

**Comment by moving-bits** [comment]
Another option would be to rename the files already existing in the target dir (instead of the copied/moved source files). Using this approach would make sure, that the set of files being copied/moved will stay consistent in terms of file names.

Renaming could be done by either appending a number or a certain extension or file name infix (e. g. "test.dat" => "test.backup.dat").

IMHO it comes down to the question which file set (source or target) we value more "important" to stay unchanged.

To keep things easier I would silently rename already existing files in the target folder and display a message on finish. Message could be something like this:
- "Some files having the same name as the source files already existed in the target folder. Renamed the existing files by adding an ".backup" infix."
- or: "There have been file name duplicates between source and target folders. Renamed the existing files by adding an ".backup" infix."
- with or without number of renamed files ("...renamed xx existing files...")

(All this assumes that you can still detect when source and target folder are identical, and do no copy/move/rename in that case.)

**Comment by eddiemuc** [comment]
> Another option would be to rename the files already existing in the target dir (instead of the copied/moved source files). Using this approach would make sure, that the set of files being copied/moved will stay consistent in terms of file names.

Haven't thought of this option, it is very valid too. Let's wait for more feedback though.

> (All this assumes that you can still detect when source and target folder are identical, and do no copy/move/rename in that case.)

This assumption is true and will stay true in all cases :-) In fact, @fm-sys already tested the case where this is happening and nothing is copied/moved (but he was not so happy with the way it is displayed in the resulting dialog and wants to change it - for which he has my blessings)

**Comment by fm-sys** [comment]
Fine for me. Asking the user would just increase complexity another time. We still have a lot of work to do with the SAF, so we shouldn't spend too long on such marginal cases. "Silently overwrite" is IMHO a no-go, we could just think about renaming the already existing files in the target folder instead of the new ones. 

Oh,I just saw, that @moving-bits have already suggested the same thing.

> To keep things easier I would silently rename already existing files in the target folder and display a message on finish. 

IMHO a message is not necessary at all, but would also be ok for me. 

> Renaming could be done by either appending a number or a certain extension or file name infix (e. g. "test.dat" => "test.backup.dat").

I would stay with the number suffix, as ".backup" is IMHO not that easy to read. Also the duplicated files do not necessarily have to have the same content so "backup" would simply be incorrect

**Comment by moving-bits** [comment]
>  Also the duplicated files do not necessarily have to have the same content so "backup" would simply be incorrect

Ah, yes, you are right. I just wanted to make it easier for the user to detect which files got renamed. And we may have files with "(1")" etc. in the source folder. - But don't know of a better solution for now.

**Comment by eddiemuc** [comment]
Wrt to the option of renaming the original file in "target". Maybe that is not such a good idea.

Reason: to get a feeling what behaviour the average user might be familiar with I checked what Windows offers in similar cases (copying/moving a folder and finding that files with same name exist in target). Windows only offers the choices to overwrite or rename the NEW (copied) file. No way to select renaming the original file in target. So maybe it is not so good to introduce some behaviour unfamiliar to the average user.

Example screenshots trying to copy (again) the content of a folder to another folder already containing all elements:
![image](https://user-images.githubusercontent.com/6909759/107142932-6513a680-6932-11eb-8847-b2848e72d000.png)

Clicking on "Für jede Datei selbst entscheiden" gives me:
![image](https://user-images.githubusercontent.com/6909759/107142956-84123880-6932-11eb-90e3-7b90bc766afc.png)

(this goes on with a list of all files in source where I can select for each file which version to keep. If both are selected then the copied version is renamed. No option to have the original file in target renamed)

**Comment by fm-sys** [comment]
Ok. Than just leave everything like it is. The chance, that this special case will occur is quite low anyway...

**Comment by fm-sys** [comment]
> There is only one little bug left. "0/50 files progressed" doesn't sound quite successful, while actually everything was done (as I selected the old folder and nothing needed to be copied). I will create a PR for this one...
> ![2021-02-06 23_51_58-cgeo Android Studio](https://user-images.githubusercontent.com/64581222/107131495-492ee700-68d7-11eb-9a8a-4797c539dc64.png)


While working on optimizing the result dialog I asked my self, if there is any good reason, why there should be an "OK" and "cancel" option. We could (or IMHO _should_) apply the folder change directly before the dialog gets displayed. Then we wouldn't have to deal with callbacks (code complexity!) and can also simplify the UI. 

@eddiemuc I guess you had implemented this, in case moving should fail. Anyway, I guess this is not clear to users anyway. To be honest, I also understood it only after looking at the code. So would it be OK for you if I remove the callback and apply the folder change directly?

Or the other way around: under which circumstances should the new folder not be applied?
| | |
|-|-|
OK | → apply
SOURCE_NOT_READABLE | → apply?
TARGET_NOT_WRITEABLE | → do not apply?
FAILURE | → apply?
ABORTED | → do not apply

**Comment by eddiemuc** [comment]
Hi @fm-sys I am perfectly fine with switching folders before copy/move and removing the option to cancel the folder change (although in general I believe that user should be able to cancel such a change until the very end)

But I would not add any black magic where success of folder change is automatically guessed by the system depending on copy/move outcome status

**Comment by fm-sys** [comment]
> But I would not add any black magic where success of folder change is automatically guessed by the system depending on copy/move outcome status 

Ok. Totally fine for me. You're right that this would cause even more confusion... 

> although in general I believe that user should be able to cancel such a change until the very end

The problem is that canceling IMHO doesn't make sense when everything has already been moved. That would result in empty folders and missing data and would lead to confusion.

Any further opinions on this?

**Comment by eddiemuc** [comment]
> > although in general I believe that user should be able to cancel such a change until the very end
> 
> The problem is that canceling IMHO doesn't make sense when everything has already been moved. That would result in empty folders and missing data and would lead to confusion.

I don't think that I am not able to foresee all possible errors or situations which can occur during the whole process and how users want to react on this. Assume that the move was already half-done due to some error, maybe the user would prefer to stay with the old folder because the "majority" of data is still there. Assume that the move fails e.g. already on the first file.

Btw, I initially planned to also include a "cancel" option for the copy/move process while it is running. The copyAll function is able to do that, only the progressbar provided by c:geo does not (yet) support a cancel button. Reason for this: I usually test by copying/moving a folder with >1000 files in it, and this can take a serious amunt of time. Users realizing that they made some clicking mistake while this is happening might want to have a cancel option.

**Comment by fm-sys** [comment]
> Btw, I initially planned to also include a "cancel" option for the copy/move process while it is running. The copyAll function is able to do that, only the progressbar provided by c:geo does not (yet) support a cancel button. 

There is a class called `SimpleDisposableHandler` which you could check out. But I don't know if it's compatible or how much work it would be to use it... If the process were cancellable, there is indeed a valid reason why a cancel should be possible. 

It would also be ok for me to keep "cancel" as long as the dialog cannot be cancelled with a click outside (which was the case, but is already changed in my local repo ;-) ).

**Comment by MagpieFourtyTwo** [comment]
> Btw, I initially planned to also include a "cancel" option for the copy/move process while it is running.

I would consider this a kind of delicate function, cause a "real" ``Cancel`` will most probably leave both folders in an inconsistent state, thus puts the user in a pretty uncomfortable situation, as he has to clean up the mess manually - and until he dd so (properly), he will in worst case not be able to use c:geo at all. Thus a "clean" ``Cancel`` should do a rollback, too - which may possibly be a bit complicated, even more if you take overwrite, rename and other things into account ... would this really be worth the effort for such a rather rare case ...?

**Comment by MagpieFourtyTwo** [comment]
WRT the wizard, we had an unexpected behavior this morning: My cachemate installed the latest nightly and on next start she got the wizard - kind of surprising, as we were pretty sure that we already had changed the folders in question ...

However, the wizard first gave the introduction (nicely done, liked it!) and finally asked for the Map files folder. So we guided the dialog to ``Internal => /Locus/mapsVector`` (she has no external SD card) and although the content of the confirmation sounded a bit weird (would be better to say ``Your map files folder has been set to xxx, yyy files and zz folders have been moved,`` or alike), things seemed to have gone right:

![1 - cgeo asked for Map folder, used previous, but reports MOVED](https://user-images.githubusercontent.com/9889015/107158992-c4eb6b00-698d-11eb-906a-16cc8cb835c8.jpg)

Nevertheless, after finishing the action and checking the settings, the Map files folder was set to ``/maps`` - strange! So we manually set it "back" to ``Internal => /Locus/mapsVector`` and checked for the folder ``/maps``, which did not even exist!

But things got even more strange with the next window, which asked for a Download folder, for GPX and stuff. Which was previously set to ``Internal/Downloads`` - thus we guided the dialog to exactly this folder again, but then the following appeared:

![2 - cgeo asked for GPX folder, got DOWNLOADS, but MOVED stuff like backups from internal-cgeo to DOWNLOADS, leaving internal data dir unchanged](https://user-images.githubusercontent.com/9889015/107159120-738fab80-698e-11eb-8fe7-6478d6c0d6cb.jpg)

Strangely enough c:geo had moved all the stuff from ``Internal => /cgeo`` (i. e. folders for Backups, fieldnotes etc.) to ``Internal => /Downloads``, which is of course completely wrong. So I manually moved the stuff back to ``Internal => /cgeo`` (where we BTW even found two new folders ``maps`` and ``gpx``). Finally I wanted to switch the settings of cgeo's base and GPX folder back to the correct values, but - surprisingly - they already were correct: The base folder was (still) set to ``/cgeo`` and the GPX showed up as ``/Download`` ...

I was pretty puzzled ... How can all this be?

**Comment by MagpieFourtyTwo** [comment]
Off topic: After upgrading this morning we got this message with ``There are problems with logging caches - we are working on it``, which in fact forced us to downgrade to use c:geo over the day. But when I now wanted to check reasons and state, I can not find a matching issue ... is there none? And, BTW. is there a (second) possibility to see this message (poss. in advance) at cgeo.org or somewhere else ...?

**Comment by moving-bits** [comment]
> Off topic: After upgrading this morning we got this message with `There are problems with logging caches - we are working on it`, which in fact forced us to downgrade to use c:geo over the day. But when I now wanted to check reasons and state, I can not find a matching issue ... is there none? And, BTW. is there a (second) possibility to see this message (poss. in advance) at cgeo.org or somewhere else ...?

This has been a temporary bug I introduced yesterday while working on viewbinding migration, see #9975. It has been fixed meanwhile & the fix will be in the next nightly version.

**Comment by Lineflyer** [comment]
> And, BTW. is there a (second) possibility to see this message (poss. in advance) at cgeo.org or somewhere else ...?

You can check on status.cgeo.org whether a notification is active.
And you could have clicked on it to get to #9775 AFAIR when I set this up.

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo 
For the maps part: no idea wh it showed /maps inbetween...
for the GPX part: Looks like for some weird reason cgeo seemed to think that the "old legacy" GPX folder was "/cgeo". This would explain the behaviour, since the wizard starts automatically a move of all files from old folder to.new one ("Download" in your case). One question to this: before SAF there were two different GPX folders configurable, one for import and one for export. Was one of the two set to /cgeo before the change? I don't know out of my head which one is used as "old folder" in case both are configured, but if that one is /cgeo this would explain the behaviour. 

Maybe assumption that automatic move is always the best option is not so true after all?

Btw:
@Lineflyer has already discovered a problem with folders resetting to default values when their configured values are not accessible at startup.

**Comment by MagpieFourtyTwo** [comment]
> One question to this: before SAF there were two different GPX folders configurable, one for import and one for export. Was one of the two set to /cgeo before the change?

Hmm, hard to say, subsequently, of course, but I would be very surprised. I would bet they were always set to ``Download``, as we exchange GPX exports via ``Downloads`` all the time and for years ...

> Maybe assumption that automatic move is always the best option is not so true after all?

As said before: IMHO the wizard should in first line just try to mirror the given situation to the new, thus leaving all the stuff where it already was for years ... cause users usually have good reasons for their files being in the current places. I in fact can not imagine not even one reason why a user should change a data folder, just because c:geo internally(!) needs a different name for the very same child.

TMALSS: I would not move a single byte, but would just try to translate the old folder name to the new notation. And nothing more. So moving stuff is a really bad thing, anyway, in my opinion.

**Comment by eddiemuc** [comment]
> > One question to this: before SAF there were two different GPX folders configurable, one for import and one for export. Was one of the two set to /cgeo before the change?
> 
> Hmm, hard to say, subsequently, of course, but I would be very surprised. I would bet they were always set to `Download`, as we exchange GPX exports via `Downloads` all the time and for years ... but can of course not put my hands in the fire for this.

Is it possible for you to retrieve the values of the following properties on the device in question (via Settings -> System -> View Properties):
"gpxExportDir", "gpxImportDir", "mapDirectory", "pref_trackfile"

(Those are, as I am sure you have guessed, the legacy property keys where the legacy folder values were stored. The values you found may have a "" prepended, this marker I use for already migrated values, but I decided to leave the old values still in the properties exactly to be able to debug situations like this)
 
> > Maybe assumption that automatic move is always the best option is not so true after all?
> 
> As said before: IMHO the wizard should in first line just try to mirror the given situation to the new one, thus leaving all the physical stuff where it already was for years ... cause users usually have good reasons for their files being in the current places. I in fact can not imagine not even one reason why a user should change a data folder, just because c:geo internally(!) needs a different name for the very same child.

I personally agree with you. In fact, in the beginning I tried to find a solution working without the whole copy/move stuff (more or less what you describe). But the move function was questioned very eagerly and no one opposed at the time, so I finally decided to implement it. I also planned to let the user select whether to move content or do nothing (and added the copy function because it was easy), questionable UX of course (that's just me :-)). Put a lot of work into it too. Anyway, I guess this is a group decision to make.

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo found the reason for the strange display of the value "/maps". Turns out this was a bug in routine for user-displayable Uris, might also be the reason for display of "/cgeo". This bug lead for some Uris to display of just the last folder name. Fixed by PR #9987

**Comment by MagpieFourtyTwo** [comment]
Yep reasonable. Even more as I just stumbled upon another strange path: Saved a logcat (cause the mentioned cachemate experienced a massively increased count of crashes during the last days), and the confirmation message said ...

![Logfile saved to logs](https://user-images.githubusercontent.com/9889015/107270780-a8ae0380-6a4b-11eb-9790-f39d2dce5dda.jpg)

**Comment by MagpieFourtyTwo** [comment]
> Is it possible for you to retrieve the values of the following properties on the device in question (via Settings -> System -> View Properties): "gpxExportDir", "gpxImportDir", "mapDirectory", "pref_trackfile"

@eddiemuc You were right, at least partly:  ;)
- gpxExportDir: /sdcard/cgeo
- gpxImportDir: /sdcard/cgeo
- mapDirectory: /storage/emulated/0/Locus/mapsVector
- pref_trackfile - setting does not exist

Nevertheless, the wizard should not have moved cgeo's base folder ... IMHO.  And it's still hard to understand why the map folder has been changed although we most definitely set it to ``/Locus/mapsVector``.
If this would have happened to an average user ... big mess.
Thus I would still prefer the defensive method, following the "Never change a running system" strategy, and just change the already set path names to the new notation - even though (and I know how hard this is) it for sure was a lot of work to implement the move stuff (I'm an SW developer myself, and I know from own experience how much every single line implemented in vain hurts :/ ).

**Comment by fm-sys** [comment]
> Thus I would still prefer the defensive method, following the "Never change a running system" strategy, and just change the already set path names to the new notation 

If we decide to move nothing when calling folder selection from the installation wizard, c:geo should at least check if the user selected the right folder as basedir (on upgrade). In addition, move should still be available when changing the folder via the settings.

Btw, what happens if the user can't find his cgeo base dir location as it is hidden in the chooser (see  #9881)? If there is no move ... big mess too. IMHO we should try to find the reason why the map folder has been changed in your case, fix the issue but leave the fine tuned move functionality in.

**Comment by MagpieFourtyTwo** [comment]
> If we decide to move nothing when calling folder selection from the installation wizard,
> c:geo should at least check if the user selected the right folder as basedir (on upgrade).

Good idea. Perhaps c:geo could remember a typical file in the legacy folder and check if this file exists with persistable path, too.
If not, the user can possibly be asked if the choice was really correct - and at this point a ``Yes (I know what I do) / No (this was an error, want to correct it, give me a second chance) / Cancel (seems to be wrong, but I first want to ask for assistance)`` would be pretty handy.  ;)

> what happens if the user can't find his cgeo base dir location as it is hidden in the chooser 

Like said in #9881 there can probably be a hint how to unhide it.

> If there is no move ... big mess too

Not really - only thing that would have been changed up to this point is the value of ``pref_<whatever>`` and c:geo could even be able to use the legacy values as fallback. Moreover, if things do not work, the user will for sure a) try to set the folder again or b) ask for assistance. And if the latter, the old folder is still stored in the legacy variables, thus this misconfiguration can easily be corrected, And in fact *way* more easy, as if the files (or even worse: just some) have been moved to a probably even unknown place. Worst case: the user did this twice. with different folders - then a) files are spread over different folders and the first is completely unknown ...

As said before: I am pretty sure that moving files is a kind of dangerous action ... esp. in a context, where most of the affected users do not even know what they are doing.

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo 
> Nevertheless, the wizard should not have moved cgeo's base folder ... IMHO. And it's still hard to understand why the map folder has been changed although we most definitely set it to /Locus/mapsVector.
If this would have happened to an average user ... big mess.

I think there is still a bug hiding in my code that some legacy file paths are somehow not recognized correctly. I am still on it.

As for the move/non-move discussion: I will accept any discussion outcome on this and not further participate in it. But in this special case, when deciding not to do it someone else would need to remove it from code, I would not have the heart myself.

**Comment by moving-bits** [comment]
IMHO we need to distinguish between three scenarios:

a) first install
   - no folders set yet, so no migration
   - just select a base folder (as part of the installation wizard), and you're done for now (in terms of SAF)

b) migration (= upgrading from an already installed older version of c:geo)
   - user needs to select one or several folders to grant SAF access to c:geo
   - if we are 100% able to detect whether the user-selected folder is the same folder as the old config, no moving of data is necessary. We could simply accept the user's choice in this case and continue with the next step.
   - if the user detects a different folder: what to do then?
     - we could warn the user - (But I would ask myself in such a case: "If you are so smart to detect that this is not the previously selected folder, then why do you bother me asking?" - Because I have no idea of (and don't care about) the technical background.)
     - we could accept the user's choice and move the data
   - and if we are not able to 100% safely detect whether the user selected the old folder, we have to move the data anyway

c) regular use via settings
   - if a user selects in the settings to change the folder, the data should be moved along with their new choice

TMALSS: I'm still convinced that moving the data according to the user's choice is the right thing to do.

If we are able to detect the previously configured folder, we explain what we do and what we expect the user to do, so we should adapt our migration wizard's text for this step, mentioning something like "we are trying to auto-detect the previously configured folder, please check and confirm, or correct when necessary" (displaying the previously configured folder along with it).

**Comment by MagpieFourtyTwo** [comment]
Sounds good.

> if the user detects a different folder: what to do then? 

We should warn him. Cause moving stuff (erroneous) can result in corrupting the base of other apps (like e. g. Locus).
Or - like we had the other day - moving cgeo's data files to download, where it will perhaps be deleted on next house keeping ...
Thus it should be absolutely and unmistakably clear that he did not chose this dir on error.

> But I would ask myself in such a case: "If you are so smart to detect that this is not the previously selected folder, then why do you bother me asking?

... cause it is not about just selecting a folder, but in fact grant allowance to use it ... this can even be understood by a not experienced user.

**Comment by rsudev** [comment]
> ... cause it is not about just selecting a folder, but in fact grant allowance to use it ... this can even be understood by a not experienced user.

Here I tend to disagree. Most 'not experienced' users I see have even a problem with understanding the concept of folders in relation to their smartphone and much less why they should suddenly grant access to something previous versions of c:geo were happily using all the time...

**Comment by MagpieFourtyTwo** [comment]
The introduction text already states that this it's not because of a change in c:geo, but in Android.

**Comment by eddiemuc** [comment]
> if we are 100% able to detect whether the user-selected folder is the same folder as the old config, no moving of data is necessary.

For the record: we **are** 100% able to detect whether the user-selected folder is the previous one. This check is already build into the copy/move routine. (It works by creating a dummy file in target folder and check whether this file is also in the source folder. This works even in mixed SAF-file-scenarios or when totally different formatted Uris Point to same physical folder)

**Comment by eddiemuc** [comment]
And by the way: why not just let the user decide what he wants to do on folder select (no copy/move, copy or move) like Implemented today in folder selection? Today it's all packed into one dialog, if this seems too much for average user then how about making a best guess, writing this into the dialog (e.g. "your content will be moved" or "...not be moved") and offer a button there titled "Advanced", opening another small dialog where user can override the system choice?

One example where users might not always want their content moved might be choosing a folder for offline maps, e.g. when they are realizing that they have different map folders for cgeo and Locus and want to unify these on their device.

**Comment by fm-sys** [comment]
IMHO the copy/move thing is understandable to users, if it's shown at the right time. So we should first check if move potentially necessary and when it makes sense ask the user if moving should really be performed. (As well in settings as in the migration dialog) 

The reason why I first didn't want that this question is asked, was because it is currently simply the wrong order. You must have chosen a folder before you can think about if content should be moved.

**Comment by eddiemuc** [comment]
> IMHO the copy/move thing is understandable to users, if it's shown at the right time. So we should first check if move potentially necessary and when it makes sense ask the user if moving should really be performed. (As well in settings as in the migration dialog) 
> 
> The reason why I first didn't want that this question is asked, was because it is currently simply the wrong order. You must have chosen a folder before you can think about if content should be moved. 

Makes a lot of sense. This would mean another dialog in the workflow, but maybe this is still the better approach.

**Comment by moving-bits** [comment]
Thanks for this discussion, I really value our striving for the most user-friendly approach!

Let me try to summarize the current proposal (as to my understanding):

(cycle for each folder)
- explanation page: what is this folder about + display its current value (either file or SAF-based)
  - (similar to what we have currently in the wizard, but added display of current folder, so that the user can check in the next step, if the preset has worked)
- open SAF file selecter with current value as preset
- on confirm check whether accepted folder equals old folder
  - if so: store result and continue silently (show conformation toast in non-migration mode)
  - if different folder: ask user whether they really want to change the folder and offer "do again", "copy", "move", "skip" options
    - "do again" opens the grant selector again
    - "skip": do not change any setting & continue to next step
    - otherwise move or copy files, show result page "xx files copied" etc.

Behavior would be the same in migration mode and in regular "change a setting" mode, with maybe the small exception that in non-migration mode we should always show a small confirmation info, even if the user has not changed the folder.

Would that be agreed upon?

**Comment by eddiemuc** [comment]
Agreed with one small addition: in non-migration mode user should have on first screen the possibility to select the folders default value.

**Comment by moving-bits** [comment]
Ok, I'll dive into it. 

@eddiemuc 
Can you give me a hint how to detect whether a selected folder is equivalent to the previous folder (even if the old one is non-SAF)? Is there some method for it, or do I have to work with some marker file as you did in `copyAllFirstPassCollectInfo`?

**Comment by eddiemuc** [comment]
> Ok, I'll dive into it. 
> 
> @eddiemuc 
> Can you give me a hint how to detect whether a selected folder is equivalent to the previous folder (even if the old one is non-SAF)? Is there some method for it, or do I have to work with some marker file as you did in `copyAllFirstPassCollectInfo`?

The only idea I had was the one with the marker file. If you have a clever alternative idea I would be very interested :-)

**Comment by eddiemuc** [comment]
Oh and thanks a lot for taking over this! I am currently swamped with work unfortunately and can't spend as much time with coding as I would like too...

**Comment by moving-bits** [comment]
The dialog itself could look like this:

![image](https://user-images.githubusercontent.com/3754370/107820464-20956a00-6d7b-11eb-94cc-98e15ef3319c.png)

I'm currently having a hard time to make this working nicely in the "uninstall + reinstall" setting for the base folder: As this is a new install / reinstall, there is no settings file read yet, and thus no "old folder" to compare against. The dialog will always pop up then, even when you selected the correct folder from the previous installation. 

Don't know yet how to detect & prevent this. (I could compare against `source == null`, which should work for a fresh install, but not in the "uninstall + reinstall" scenario. Also we do have the `PersistableFolder.LEGACY_CGEO_PUBLIC_ROOT` I could use as "educated guess", but that won't work under API 30 any longer AFAICS.)

**Comment by fm-sys** [comment]
What about checking the default directorys backup snd logfiles, which are AFAIK always there. But if there's no folder migrating from, the move stuff should however anyway be skipped ccompletely. (fresh install - > no data migration) 

Oh, and what does "skip folder" mean? Is the folder not saved or only no data moved. This is not clear from the wording...

**Comment by moving-bits** [comment]
> What about checking the default directorys backup snd logfiles, 

You mean: Checking whether subfolders "backup" or "logfiles" exist in the selected target folder?

By that I could indeed detect whether the user has selected the existing base folder, even in the "uninstall + reinstall" scenario (at least if the user has done at least one backup, which would be helpful in that scenario anyway), but it would not work if the user chooses a different folder, which would be the actual use-case for copy/move actions, as I could not detect the actual previous base folder to copy/move from.

> But if there's no folder migrating from, the move stuff should however anyway be skipped ccompletely.

Hm, I guess I cannot distinguish between a fresh install and a reinstall scenario with new folder chosen, so I will always have to skip that menu.

> Oh, and what does "skip folder" mean? Is the folder not saved or only no data moved. This is not clear from the wording...

"Skip folder" means "do not store any grants for this folder now". What naming would you suggest? Maybe "revert"? Or "skip, do not store"?

**Comment by fm-sys** [comment]
> Hm, I guess I cannot distinguish between a fresh install and a reinstall scenario with new folder chosen, so I will always have to skip that menu.

👍



> > Oh, and what does "skip folder" mean? Is the folder not saved or only no data moved. This is not clear from the wording...
> 
> "Skip folder" means "do not store any grants for this folder now". What naming would you suggest? Maybe "revert"? Or "skip, do not store"?

Honestly, I haven't thought about it yet because I really didn't know what it was for. But wouldn't a simple cancel action button serve your purpose?

**Comment by eddiemuc** [comment]
> The dialog itself could look like this:
> 
> ![image](https://user-images.githubusercontent.com/3754370/107820464-20956a00-6d7b-11eb-94cc-98e15ef3319c.png)
> 
> I'm currently having a hard time to make this working nicely in the "uninstall + reinstall" setting for the base folder: As this is a new install / reinstall, there is no settings file read yet, and thus no "old folder" to compare against. The dialog will always pop up then, even when you selected the correct folder from the previous installation. 

I really think it is fine to let this one be. We can't handle every case....

I am missing the option "select folder but do neither move or copy any content".

**Comment by moving-bits** [comment]
Ok, changed a few things:
- If no previous folder is known (first install / reinstall scenarios) skip the dialog and continue with `CopyChoice.DO_NOTHING`
- Move the "skip" option to a "Cancel" button
- Added the "Select folder, no further action" item

Remark: For "base folder" there will never be a known previous folder, as there is no equivalent in the settings. Shall I add our `LEGACY_CGEO_PUBLIC_ROOT` as temporary "educated best guess" to `Settings.getLegacyValue()`?


Dialog looks like this now:

![image](https://user-images.githubusercontent.com/3754370/107845620-1310cc00-6ddd-11eb-8933-5de3135bc3b7.png)

**Comment by moving-bits** [comment]
Regarding the dialog: Alternatively we could use this approach (distinguishes more clearly between retry and select action):

![image](https://user-images.githubusercontent.com/3754370/107845839-e958a480-6dde-11eb-99f5-dfee105eecda.png)

**Comment by eddiemuc** [comment]
> Remark: For "base folder" there will never be a known previous folder, as there is no equivalent in the settings. Shall I add our `LEGACY_CGEO_PUBLIC_ROOT` as temporary "educated best guess" to `Settings.getLegacyValue()`?

On the contrary, base will always have this folder as current folder if no folder was explicitly set yet (if you call `getFolder()` on the base folder on a fresh c:geo installation you will get the legacy folder from below in return). This is the magic of folder default values. I would not mix this behaviour with legacy migration. The legacy migration is just for handling explicit folder settings done by the user in previous c: geo versions.

Same goes for all folders btw: if for a folder no explicit value was set yet, they have a value nevertheless which is a subfolder of base.

**Comment by eddiemuc** [comment]
Have to add to the above: base will have this value if no explicit value was set yet by the user **AND** the folder is still accessible by c:geo. If the later is not the case then there is a second fallback in documents/cgeo. If this is not accessible either, the final fallback would be a directory in c:geo private storage. 

In any case, base will always have a current folder value also if not explicitly set

**Comment by moving-bits** [comment]
Hm, I'm entering the migration routines the same way for `PersistableFolder.BASE` and the others, e. g. `PersistableFolder.OFFLINE_MAPS`. After some steps I return to `onActivityResult` which leads me to `handleResultPersistableFolderSelection`. Here's where the magic happens:

```
final Folder sourceFolder = Folder.fromConfig(Settings.getPersistableFolder(folder));
if (runningIntentData.copyChoice == CopyChoice.ASK_IF_DIFFERENT && sourceFolder != null && !FolderUtils.get().foldersAreEqual(sourceFolder, Folder.fromDocumentUri(uri))) { ... }
```

Do I need to call something different to get the current value for source?

**Comment by fm-sys** [comment]
> Regarding the dialog: Alternatively we could use this approach (distinguishes more clearly between retry and select action):
> 
> ![image](https://user-images.githubusercontent.com/3754370/107845839-e958a480-6dde-11eb-99f5-dfee105eecda.png)

If we want to distinguish, we should probably use a real button inside the layout or move it to the actions at the bottom as well (first option preferred)

**Comment by eddiemuc** [comment]
To get the current folder set, call `PersistableFolder.BASE.getFolder()`. Do not use `Settings`.

I designed the system such that only the `PersistableFolder` class and `ContentStorage` need to deal with `Setting`. No one else shall call Settings for folders directly (I think/hope I commented the settings methods accordingly, unfortunately Java does not offer a way to restrict calls from other classes by compiler)

**Comment by eddiemuc** [comment]
Checked Settings, and I did not put this in the comments  unfortunately. this is my fault, sorry for that. Will change that next time I am coding.

**Comment by moving-bits** [comment]
Thanks, that helps a lot. I'll adapt the code accordingly. And I've added comments to the special methods in `Settings`.

**Comment by moving-bits** [comment]
In the "fresh install" scenario I have an error in the log, and the base folder points to the internal fallback folder, although at the time migration is called storage permission has been granted, so the default should/can point to the public storage folders:

```
2021-02-13 10:51:06.588 2583-2611/? E/LoadDocStackTask: Failed to build document stack for uri: content://com.android.externalstorage.documents/document/user%3A0%2Fcgeo.geocaching%2Ffiles%2Fpublic%2FBASE
    java.io.FileNotFoundException: No root for user
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.os.ParcelableException.readFromParcel(ParcelableException.java:56)
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:82)
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:79)
        at android.os.Parcel.readParcelable(Parcel.java:2973)
        at android.os.Parcel.createException(Parcel.java:2066)
        at android.os.Parcel.readException(Parcel.java:2039)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:658)
        at android.content.ContentProviderClient.call(ContentProviderClient.java:558)
        at android.content.ContentResolver.call(ContentResolver.java:2032)
        at android.provider.DocumentsContract.findDocumentPath(DocumentsContract.java:1590)
        at com.android.documentsui.DocumentsAccess$RuntimeDocumentAccess.findDocumentPath(DocumentsAccess.java:130)
        at com.android.documentsui.LoadDocStackTask.run(LoadDocStackTask.java:75)
        at com.android.documentsui.LoadDocStackTask.run(LoadDocStackTask.java:43)
        at com.android.documentsui.base.CheckedTask.doInBackground(CheckedTask.java:65)
        at android.os.AsyncTask$3.call(AsyncTask.java:378)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at com.android.documentsui.ProviderExecutor.run(ProviderExecutor.java:104)
```

and later during folder migration (own log message in `handleResultPersistableFolderSelection()`):
```
change folder: (same=false)
source=/data/user/0/cgeo.geocaching/files/public/BASE[FILE#2:p-file:///data/user/0/cgeo.geocaching/files::/public/BASE]
target=content://com.android.externalstorage.documents/tree/primary%3Acgeo
```

Is there a way to make `PersistableFolder` make reset/reevaluate the defaults?

**Comment by eddiemuc** [comment]
> Is there a way to make PersistableFolder reset/reevaluate the defaults?

This scenario I haven't foreseen, so currently there is no such possibility. It should however be easy to add.

Currently the default folder access evaluation happens in constructor of ContentStorage (meaning: at startup) for all folders. I assume it is possible to extract this in a method "reevaluateDefaults()" or something similar, call this in constructor and also from the outside.

**Comment by moving-bits** [comment]
Thanks, that's exactly what I've come up with in the meantime and what my compiler currently is crunching at :-)

**Comment by moving-bits** [comment]
Ok, seems I have a working solution now for either install, migration and change settings scenarios. The last one looks like this now:

|folder is currently set to default value:|folder is currently set to a user-defined value:|
|---|---|
|![image](https://user-images.githubusercontent.com/3754370/107848998-72300a00-6df8-11eb-96a8-d0e57aca76f8.png)|![image](https://user-images.githubusercontent.com/3754370/107849006-84aa4380-6df8-11eb-849d-ebd7f8c1ae39.png)|

When you select a different folder than the current you'll get this dialog:
![image](https://user-images.githubusercontent.com/3754370/107849032-bcb18680-6df8-11eb-89be-46ef4e107858.png)

The first two examples will only be shown when coming from settings, but never in install/migration mode, whereas the last dialog may also appear in migration mode.

**Comment by moving-bits** [comment]
One small inconsistency, though:
The last dialog uses "Back" as neutral option and "Cancel" as negative, whereas the first two dialogs use "Cancel" as neutral and "Use default" as negative. I guess it was done this way to have the two "select a folder" options closely together, but I would propose to swap this: use "Use default" as neutral option and "cancel" as negative option. This would be more consistent to Android-standards (and additionally I don't know how screen readers are interacting with "neutral", "positive" and "negative" options).

**Comment by eddiemuc** [comment]
I like it!

And I would be fine with swapping neutral and negative option too. You are right, I did it like this to have "select default" and "choose folder" closer together. I don't mind if this changes though.

**Comment by moving-bits** [comment]
Ok, I've created #10005, which will also close this issue. If we have more ideas for workflow improvements here, I suggest to open a new issue.

... unless we want to create an item having 100 comments, which we are close to :-)
